src/main/java/org/spongepowered/api/command/parameter/CommandContext.java

```diff
@@ -176,24 +177,69 @@
          * Creates a snapshot of this context and returns a state object that can
          * be used to restore the context to this state.
          *
+         * <p>If this is already a {@link Transaction}, this does not start a
+         * transaction, it simply returns the current one.</p>
+         *
          * @return The state.
          */
-        State getState();
+        Transaction startTransaction();
 
         /**
-         * Uses a previous snapshot of the context and restores it to that state.
+         * Creates a {@link CommandContext}.
          *
-         * @param state The state obtained from {@link #getState()}
+         * @param input The input argument string.
+         * @return The context
          */
-        void setState(State state);
+        CommandContext build(String input);
 
         /**
-         * An immutable snapshot of a {@link CommandContext.Builder}.
-         *
-         * <p>No assumptions should be made about the form of this state object,
-         * it is not defined in the API and may change at any time.</p>
+         * Starts a {@link Transaction} on the context builder.
          */
-        interface State {}
+        interface Transaction extends CommandContext.Builder {
+
+            /**
+             * Returns whether the transaction is active.
+             *
+             * @return If is active.
+             */
+            boolean isActive();
+
+            /**
+             * Commits this {@link Transaction} to the associated builder.
+             *
+             * <p>This {@link Transaction} becomes inactive after this is called
+             * and can no longer be used.</p>
+             */
+            void commit();
+
+            /**
+             * Rolls back the changes to the associated builder.
+             *
+             * <p>This {@link Transaction} becomes inactive after this is called
+             * and can no longer be used.</p>
+             */
+            void rollback();
+
+            /**
+             * A {@link Transaction} cannot be started from within a transaction.
+             *
+             * @return The current transaction.
+             */
+            @Override
+            @Deprecated
```

Comment:
Why is new API deprecated?

Reply:
The `Transaction` was intended to be a replacement for `CommandContext.Builder` when sending it through elements. Building the builder is part of that interface, but transactions shouldn't be built. Same with starting a transaction from a transaction.

I think I'll ditch that notion and just have a Transaction as a stack frame like in causes. It'll minimise changes I have to make elsewhere, but just makes the context builder impl more complicated.

That's what implementation is for though!